### PR TITLE
[DRAFT] Stencil and culling support

### PR DESCRIPTION
**Pull request by nokola**

### CODE REVIEW

**Comment by nokola** [comment]
I'm working on the last piece needed for my vector library - stencil support.
However, I'm not sure how to get `glStencilOpSeparate` to work. I know on Windows there's GetProcAddress(), but not sure how that works with sokol/miniquad.

Created this draft pull request to discuss if you have any ideas how to get glStencilOpSeparate to compile (see TODO in source)

Note: the functions in the PR are not final, I plan to have these ones working instead:
```
glStencilMaskSeparate(...);
glStencilFuncSeparate(...);
glStencilOpSeparate(...);
glCullFace(GL_BACK);
glFrontFace(GL_CCW);

```

File: src/graphics.rs [hunk 1]
```diff
@@ -1002,13 +1002,17 @@ impl From<BlendFactor> for GLenum {
...
+    pub front_face_order: FrontFaceOrder, // TODO: UNUSED, should call glFrontFace
     pub depth_test: Comparison,
     pub depth_write: bool,
     pub depth_write_offset: Option<(f32, f32)>,
     pub color_blend: Option<BlendState>,
     pub color_write: ColorMask,
+    // TODO: support:
+    // glStencilMaskSeparate(...);
+    // glStencilFuncSeparate(...);
+    // glStencilOpSeparate(...);
```
Comment (nokola) [inline src/graphics.rs:1015]: @not-fl3 do you have any preferences or suggestions on how to structure the PipelineParams to support the above 3 functions? I'm still thinking about it.

File: native/sapp-windows/external/sokol/sokol_app.h [hunk 2]
```diff
@@ -3222,6 +3223,9 @@ typedef void  (GL_APIENTRY *PFN_glStencilFuncSeparate)(GLenum face, GLenum func,
 static PFN_glStencilFuncSeparate _sapp_glStencilFuncSeparate;
 typedef void  (GL_APIENTRY *PFN_glStencilOpSeparate)(GLenum face, GLenum sfail, GLenum dpfail, GLenum dppass);
 static PFN_glStencilOpSeparate _sapp_glStencilOpSeparate;
+void glStencilOpSeparate(GLenum face, GLenum sfail, GLenum dpfail, GLenum dppass) {
```
Comment (nokola) [inline native/sapp-windows/external/sokol/sokol_app.h:3226]: Question: why wasn't this function already auto-generated in sokol_app.h - how does the generation work?

Wondering if I should add it manually or there's some automated way that you use to generate some of the functions in this file.

Comment (not-fl3) [reply native/sapp-windows/external/sokol/sokol_app.h:3226]: All that functions were hand-written, just like you did for `glStencilOpSeparate`. 
In C function pointers was called directly with macroses, but for rust I did not found anything smarter than just handwrite a few function declarations.

File: src/graphics.rs (hunk 1 above)
Comment (not-fl3) [reply src/graphics.rs:1015]: This is how it looks in sokol_gfx, just for reference:

```C
typedef struct sg_pipeline_desc {
..
    sg_depth_stencil_state depth_stencil;
..
}
```

```C
typedef struct sg_depth_stencil_state {
    sg_stencil_state stencil_front;
    sg_stencil_state stencil_back;
    sg_compare_func depth_compare_func;
    bool depth_write_enabled;
    bool stencil_enabled;
    uint8_t stencil_read_mask;
    uint8_t stencil_write_mask;
    uint8_t stencil_ref;
} sg_depth_stencil_state;
```

Maybe we can do similar to `color_blend`? 
Something like
```rust
...
stencil: Option<StencilState>
...

```rust
struct BlendState {
    front: Foo,
    back: Foo,
    read_mask: Mask,
    write_mask: Mask,
    ref: Foo
}
```

Not sure, just first idea :)

File: native/sapp-windows/external/sokol/sokol_app.h (hunk 2 above)
Comment (nokola) [reply native/sapp-windows/external/sokol/sokol_app.h:3226]: good to know, thanks!

File: src/graphics.rs (hunk 1 above)
Comment (nokola) [reply src/graphics.rs:1114]: Created new PR with just the options.

**Comment by nokola** [comment]
Abandoning this request because changes are very hard for me to manage after the force-push from not-fl3:master. The commits in GitHub now how 20 or so commits as "difference" that I already merged before I started the PR. 
Perhaps I don't know how to use a git tool or something like that but would rather just start a new PR.

**Pull request closed without merging**